pymc_marketing/mmm/utility.py

```diff
@@ -0,0 +1,546 @@
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+"""
+Utility functions for Bayesian optimization.
+
+Key Concepts:
+-------------
+
+- **Samples**:
+    A PyTensor tensor variable (`pt.TensorVariable`) representing samples drawn from the posterior
+    distributions of the model outputs. These samples capture the uncertainty in the model predictions
+    and are essential for computing expected utilities and risk measures in Bayesian optimization.
+
+- **Budgets**:
+    A PyTensor tensor variable representing a set of monetary budgets allocated to different assets,
+    investments, or channels. Each element corresponds to the budget for a specific option in the
+    optimization process.
+"""
+
+from collections.abc import Callable
+
+import pytensor.tensor as pt
+
+UtilityFunction = Callable[[pt.TensorVariable, pt.TensorVariable], float]
+
+
+def _compute_quantile(x: pt.TensorVariable, q: float) -> pt.TensorVariable:
+    """
+    Compute the quantile of a PyTensor tensor variable.
+
+    Parameters
+    ----------
+    x : pt.TensorVariable
+        A 1D PyTensor tensor variable containing samples.
+    q : float
+        The quantile to compute, between 0 and 1.
+
+    Returns
+    -------
+    pt.TensorVariable
+        The quantile value.
+    """
+    sorted_x = pt.sort(x)
+    n = x.shape[0]
+    idx = q * (n - 1)
+    idx_floor = pt.floor(idx).astype("int64")
+    idx_ceil = pt.ceil(idx).astype("int64")
+    weight = idx - idx_floor
+    return (1 - weight) * sorted_x[idx_floor] + weight * sorted_x[idx_ceil]
+
+
+def average_response(
+    samples: pt.TensorVariable, budgets: pt.TensorVariable
+) -> pt.TensorVariable:
+    """Compute the average response of the posterior predictive distribution."""
+    return pt.mean(samples)
+
+
+def tail_distance(confidence_level: float = 0.75) -> UtilityFunction:
+    R"""Calculate the absolute distance between the mean and the quantiles.
+
+    It is a simple and interpretable metric that can be used to assess the risk.
+
+    The tail distance is calculated as:
+
+    .. math::
+        Tail\\ Distance = |Q_{(1 - \\alpha)} - \\mu| + |\\mu - Q_{\\alpha}|
+
+    where:
+        - :math:`\\mu` is the mean of the sample returns.
+        - :math:`Q_{(1 - \\alpha)}` is the quantile at the specified confidence level.
+        - :math:`Q_{\\alpha}` is the quantile at the specified confidence level.
+
+    Parameters
+    ----------
+    confidence_level : float, optional
+        Confidence level for the quantiles (default is 0.75).
+
+    Returns
+    -------
+    UtilityFunction
+        A function that calculates the tail distance metric given samples and budgets.
+    """
+    if not 0 < confidence_level < 1:
+        raise ValueError("Confidence level must be between 0 and 1.")
+
+    def _tail_distance(
+        samples: pt.TensorVariable, budgets: pt.TensorVariable
```

Comment:
we are not using the `budgets` variable, so we should remove it right?

Reply:
Replied before!